### PR TITLE
docs: complete interface & use inheritdoc

### PR DESCRIPTION
**Pull request by sakulstra**
moving structs & docs to the interface so it's easier to consume

### CODE REVIEW

**Comment by sakulstra** [comment]
@kartojal / @miguelmtzinf also noticed that
https://github.com/aave/aave-v3-core/blob/master/contracts/interfaces/IReserveInterestRateStrategy.sol#L11
only defines a super small subset of
https://github.com/aave/aave-v3-core/blob/master/contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol#L21

So again ppl relying on this usually completely redefine this interface (at least we do). Might also make sense to apply a similar change there? I'm OOO now, so can't do this.

File: contracts/misc/AaveProtocolDataProvider.sol [hunk 1]
```diff
@@ -25,22 +25,13 @@ contract AaveProtocolDataProvider is IPoolDataProvider {
...
   address constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
 
-  struct TokenData {
-    string symbol;
-    address tokenAddress;
-  }
-
   IPoolAddressesProvider public immutable ADDRESSES_PROVIDER;
 
   constructor(IPoolAddressesProvider addressesProvider) {
```
Comment (miguelmtzinf) [inline contracts/misc/AaveProtocolDataProvider.sol:30]: Could you add docs to the constructor, just for the sake of completeness

File: contracts/misc/AaveProtocolDataProvider.sol [hunk 2]
```diff
@@ -25,22 +25,13 @@ contract AaveProtocolDataProvider is IPoolDataProvider {
...
     ADDRESSES_PROVIDER = addressesProvider;
   }
 
-  /**
-   * @notice Returns the list of the existing reserves in the pool.
-   * @dev Handling MKR and ETH in a different way since they do not have standard `symbol` functions.
-   * @return The list of reserves, pairs of symbols and addresses
-   */
+  /// @inheritdoc IPoolDataProvider
   function getAllReservesTokens() external view returns (TokenData[] memory) {
```
Comment (miguelmtzinf) [inline contracts/misc/AaveProtocolDataProvider.sol:35]: Missing `override` on all these functions

**Comment by miguelmtzinf** [comment]
I was tempted to do these changes on the Interest Rate contracts back in the days but declined that idea because an asset could have a custom interest rate strategy, and the only one function needed for that is already part of the `IReserveInterestRateStrategy`.
Therefore, I am in favor of leaving the `IReserveInterestRateStrategy` as is, and create an `IDefaultInterestRateStrategy` that inherits from the other that contains all the functions of `DefaultInterestRateStrategy`

File: contracts/misc/AaveProtocolDataProvider.sol [hunk 3]
```diff
@@ -25,23 +25,19 @@ contract AaveProtocolDataProvider is IPoolDataProvider {
...
-    string symbol;
-    address tokenAddress;
-  }
-
+  /// @inheritdoc IPoolDataProvider
   IPoolAddressesProvider public immutable ADDRESSES_PROVIDER;
 
+  /**
+   * @notice Constructor
+   * @param addressesProvider address for the PoolAddressesProvider contract.
```
Comment (miguelmtzinf) [inline contracts/misc/AaveProtocolDataProvider.sol:33]:
```suggestion
   * @param addressesProvider The address for the PoolAddressesProvider contract
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  /**
+   * @notice Returns the stable rate excess offset
+   * @dev An additional premium applied to the stable when stable debt > OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO
+   * @return The stable rate excess offset
+   */
+  function getStableRateExcessOffset() external view returns (uint256);
+
+  /**
+   * @notice Returns the base stable borrow rate
+   * @return The base stable borrow rate
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:81]:
```suggestion
   * @return The base stable borrow rate, expressed in ray
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+   * @notice Returns the stable rate slope below optimal usage ratio
+   * @dev Its the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
+   * @return The stable rate slope
+   */
+  function getStableRateSlope1() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate slope above optimal usage ratio
+   * @dev Its the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
+   * @return The stable rate slope
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:68]:
```suggestion
   * @return The stable rate slope, expressed in ray
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+   * @notice Returns the variable rate slope above optimal usage ratio
+   * @dev Its the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
+   * @return The variable rate slope
+   */
+  function getVariableRateSlope2() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate slope below optimal usage ratio
+   * @dev Its the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
+   * @return The stable rate slope
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:61]:
```suggestion
   * @return The stable rate slope, expressed in ray
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  /**
+   * @dev Returns the address for the PoolAddressesProvider contract.
+   */
+  function ADDRESSES_PROVIDER() external view returns (IPoolAddressesProvider);
+
+  /**
+   * @notice Returns the variable rate slope below optimal usage ratio
+   * @dev Its the variable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
+   * @return The variable rate slope
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:47]:
```suggestion
   * @return The variable rate slope, expressed in ray
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import {IReserveInterestRateStrategy} from './IReserveInterestRateStrategy.sol';
+import {IPoolAddressesProvider} from './IPoolAddressesProvider.sol';
+
+/**
+ * @title IDefaultInterestRateStrategy
+ * @author Aave
+ * @notice Default Interface for the calculation of the interest rates
+ */
+interface IDefaultInterestRateStrategy is IReserveInterestRateStrategy {
+  /**
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:13]: Redo the docs of these functions with proper @param tags and changing the @dev to @notice, for consistency

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.0;
+
+import {IReserveInterestRateStrategy} from './IReserveInterestRateStrategy.sol';
+import {IPoolAddressesProvider} from './IPoolAddressesProvider.sol';
+
+/**
+ * @title IDefaultInterestRateStrategy
+ * @author Aave
+ * @notice Default Interface for the calculation of the interest rates
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:10]:
```suggestion
 * @notice Defines the basic interface of the DefaultReserveInterestRateStrategy
```

File: contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol [hunk 10]
```diff
@@ -5,9 +5,10 @@ import {IERC20} from '../../dependencies/openzeppelin/contracts/IERC20.sol';
 import {WadRayMath} from '../libraries/math/WadRayMath.sol';
 import {PercentageMath} from '../libraries/math/PercentageMath.sol';
 import {DataTypes} from '../libraries/types/DataTypes.sol';
-import {IReserveInterestRateStrategy} from '../../interfaces/IReserveInterestRateStrategy.sol';
 import {IPoolAddressesProvider} from '../../interfaces/IPoolAddressesProvider.sol';
 import {Errors} from '../libraries/helpers/Errors.sol';
+import {IDefaultInterestRateStrategy} from "../../interfaces/IDefaultInterestRateStrategy.sol";
+import {IReserveInterestRateStrategy} from "../../interfaces/IReserveInterestRateStrategy.sol";
```
Comment (miguelmtzinf) [inline contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol:11]: seems not needed

File: contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol [hunk 11]
```diff
@@ -5,9 +5,10 @@ import {IERC20} from '../../dependencies/openzeppelin/contracts/IERC20.sol';
 import {WadRayMath} from '../libraries/math/WadRayMath.sol';
 import {PercentageMath} from '../libraries/math/PercentageMath.sol';
 import {DataTypes} from '../libraries/types/DataTypes.sol';
-import {IReserveInterestRateStrategy} from '../../interfaces/IReserveInterestRateStrategy.sol';
 import {IPoolAddressesProvider} from '../../interfaces/IPoolAddressesProvider.sol';
 import {Errors} from '../libraries/helpers/Errors.sol';
+import {IDefaultInterestRateStrategy} from "../../interfaces/IDefaultInterestRateStrategy.sol";
```
Comment (miguelmtzinf) [inline contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol:10]: move up this import, right before the `IPoolAddressesProvider` import please

File: contracts/mocks/tests/MockReserveInterestRateStrategy.sol [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
 pragma solidity 0.8.10;
 
-import {IReserveInterestRateStrategy} from '../../interfaces/IReserveInterestRateStrategy.sol';
 import {IPoolAddressesProvider} from '../../interfaces/IPoolAddressesProvider.sol';
 import {WadRayMath} from '../../protocol/libraries/math/WadRayMath.sol';
 import {DataTypes} from '../../protocol/libraries/types/DataTypes.sol';
+import {IDefaultInterestRateStrategy} from "../../interfaces/IDefaultInterestRateStrategy.sol";
 
-contract MockReserveInterestRateStrategy is IReserveInterestRateStrategy {
+contract MockReserveInterestRateStrategy is IDefaultInterestRateStrategy {
```
Comment (miguelmtzinf) [inline contracts/mocks/tests/MockReserveInterestRateStrategy.sol:9]: Would be nice to check if a `MockReserveInterestRateStrategy` that only implements `IReserveInterestRateStrategy` is enough to satisfy the test suite. if so, I would be in favor of pruning this mock contract.

File: contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol (hunk 10 above)
Comment (sendra) [reply contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol:11]: This is needed for the inherit docs

File: contracts/mocks/tests/MockReserveInterestRateStrategy.sol (hunk 12 above)
Comment (sendra) [reply contracts/mocks/tests/MockReserveInterestRateStrategy.sol:9]: It is enough (tested it) but I think its ok to leave the full mock, as for what i have seen, all interest rate strategies use all the methods, so could be used for someone

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+import {IPoolAddressesProvider} from './IPoolAddressesProvider.sol';
+
+/**
+ * @title IDefaultInterestRateStrategy
+ * @author Aave
+ * @notice Defines the basic interface of the DefaultReserveInterestRateStrategy
+ */
+interface IDefaultInterestRateStrategy is IReserveInterestRateStrategy {
+  /**
+   * @notice This constant represents the usage ratio at which the pool aims to obtain most competitive borrow rates.
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:14]:
```suggestion
   * @notice Returns the optimal usage ratio, which represents the usage ratio at which the pool aims to obtain most competitive borrow rates.
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The optimal usage ratio, expressed in ray.
+   */
+  function OPTIMAL_USAGE_RATIO() external view returns (uint256);
+
+  /**
+   * @notice This constant represents the optimal stable debt to total debt ratio of the reserve.
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:20]:
```suggestion
   * @notice Returns the optimal stable to total debt ratio of the reserve.
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The optimal stable to total debt ratio, expressed in ray.
+   */
+  function OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO() external view returns (uint256);
+
+  /**
+   * @notice This constant represents the excess usage ratio above the optimal. It's always equal to
+   * 1-optimal usage ratio. Added as a constant here for gas optimizations.
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:27]:
```suggestion
   * @notice Returns the excess usage ratio above the optimal.
   * @dev It's always equal to 1-optimal usage ratio (added as a constant for gas optimizations)
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The max excess usage ratio, expressed in ray.
+   */
+  function MAX_EXCESS_USAGE_RATIO() external view returns (uint256);
+
+  /**
+   * @notice This constant represents the excess stable debt ratio above the optimal. It's always equal to
+   * 1-optimal stable to total debt ratio. Added as a constant here for gas optimizations.
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:34]:
```suggestion
   * @notice Returns the excess stable debt ratio above the optimal.
   * @dev It's always equal to 1-optimal stable to total debt ratio (added as constant for gas optimizations)
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The max excess stable to total debt ratio, expressed in ray.
+   */
+  function MAX_EXCESS_STABLE_TO_TOTAL_DEBT_RATIO() external view returns (uint256);
+
+  /**
+   * @notice Returns the address for the PoolAddressesProvider contract.
+   * @return The PoolAddressesProvider contract address.
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:41]:
```suggestion
   * @notice Returns the address for the PoolAddressesProvider
   * @return The address of the PoolAddressesProvider contract
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+   */
+  function ADDRESSES_PROVIDER() external view returns (IPoolAddressesProvider);
+
+  /**
+   * @notice Returns the variable rate slope below optimal usage ratio
+   * @dev Its the variable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:47]:
```suggestion
   * @dev It is the variable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The variable rate slope, expressed in ray
+   */
+  function getVariableRateSlope1() external view returns (uint256);
+
+  /**
+   * @notice Returns the variable rate slope above optimal usage ratio
+   * @dev Its the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:54]:
```suggestion
   * @dev It is the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The variable rate slope, expressed in ray
+   */
+  function getVariableRateSlope2() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate slope below optimal usage ratio
+   * @dev Its the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:61]:
```suggestion
   * @dev It is the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+  /**
+   * @notice Returns the stable rate slope above optimal usage ratio
+   * @dev Its the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
+   * @return The stable rate slope, expressed in ray
+   */
+  function getStableRateSlope2() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate excess offset
+   * @dev An additional premium applied to the stable when stable debt > OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:75]:
```suggestion
   * @dev It is an additional premium applied to the stable when stable debt > OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The stable rate excess offset
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:76]:
```suggestion
   * @return The stable rate excess offset, expressed in ray
```

File: contracts/interfaces/IPoolDataProvider.sol [hunk 23]
```diff
@@ -1,7 +1,123 @@
 // SPDX-License-Identifier: AGPL-3.0
 pragma solidity ^0.8.0;
 
+import { IPoolAddressesProvider } from './IPoolAddressesProvider.sol';
+
 interface IPoolDataProvider {
```
Comment (miguelmtzinf) [inline contracts/interfaces/IPoolDataProvider.sol:6]: Missing docs here

Comment (miguelmtzinf) [inline contracts/interfaces/IPoolDataProvider.sol:6]:
```suggestion
  /**
   * @title IPoolDataProvider
   * @author Aave
   * @notice Defines the basic interface of a PoolDataProvider
   */
interface IPoolDataProvider {
```

File: contracts/interfaces/IPoolDataProvider.sol [hunk 24]
```diff
@@ -1,7 +1,123 @@
...
+      uint256 reserveFactor,
+      bool usageAsCollateralEnabled,
+      bool borrowingEnabled,
+      bool stableBorrowRateEnabled,
+      bool isActive,
+      bool isFrozen
+    );
+
+  /**
+   * Returns the efficiency mode category of the reserve
```
Comment (miguelmtzinf) [inline contracts/interfaces/IPoolDataProvider.sol:63]:
```suggestion
   * @notice Returns the efficiency mode category of the reserve
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+  /**
+   * @notice Returns the address of the PoolAddressesProvider
+   * @return The address of the PoolAddressesProvider contract
+   */
+  function ADDRESSES_PROVIDER() external view returns (IPoolAddressesProvider);
+
+  /**
+   * @notice Returns the variable rate slope below optimal usage ratio
+   * @dev It is the variable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:47]:
```suggestion
   * @dev It's the variable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The variable rate slope, expressed in ray
+   */
+  function getVariableRateSlope1() external view returns (uint256);
+
+  /**
+   * @notice Returns the variable rate slope above optimal usage ratio
+   * @dev It is the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:54]:
```suggestion
   * @dev It's the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The variable rate slope, expressed in ray
+   */
+  function getVariableRateSlope2() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate slope below optimal usage ratio
+   * @dev It is the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:61]:
```suggestion
   * @dev It's the stable rate when usage ratio > 0 and <= OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The stable rate slope, expressed in ray
+   */
+  function getStableRateSlope1() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate slope above optimal usage ratio
+   * @dev It is the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:68]:
```suggestion
   * @dev It's the variable rate when usage ratio > OPTIMAL_USAGE_RATIO
```

File: contracts/interfaces/IDefaultInterestRateStrategy.sol [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+   * @return The stable rate slope, expressed in ray
+   */
+  function getStableRateSlope2() external view returns (uint256);
+
+  /**
+   * @notice Returns the stable rate excess offset
+   * @dev It is an additional premium applied to the stable when stable debt > OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO
```
Comment (miguelmtzinf) [inline contracts/interfaces/IDefaultInterestRateStrategy.sol:75]:
```suggestion
   * @dev It's an additional premium applied to the stable when stable debt > OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO
```

File: contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol (hunk 10 above)
Comment (sendra) [reply contracts/protocol/pool/DefaultReserveInterestRateStrategy.sol:9]: https://github.com/aave/aave-v3-core/pull/763/files#diff-1f7acd1ca2faeee390ca9f1bde1da92aa4397a3307a8f43f2b651f8e01dc5a0cR158

**Review by miguelmtzinf — APPROVED** [review]
LGTM
Thanks @sakulstra and @sendra !

**Comment by miguelmtzinf** [comment]
Merging to an intermediate branch so I can fix the format of some contracts.

**Pull request merged**